// quart/Quart.js
const http2 = require('http2');
const Router = require('./lib/Router');
const SafeError = require('./lib/SafeError');

/**
 * Quart web server
 */
class Quart {
  /**
   * Creates a new Quart instance
   * @param {object} [options] - Options passed to createSecureServer such as `cert` and `key`
   * @param {Http2SecureServer} [instance] - Optional http2 instance to be used instead of creating a brand new one
   */
  constructor(options, server) {
    // Store options
    this.options = options;

    // Initialize server or reuse from args
    if (!server) this.server = http2.createSecureServer(options);
    else this.server = server;

    // Create a Router instance
    this.router = new Router();

    // Bind individual streams to Router
    this.server.on('stream', (stream, headers) => this.router.routeStream(stream, headers));
  }

  get SafeError() {
    return SafeError;
  }

  static get SafeError() {
    return SafeError;
  }

  /**
   * Listen on a path
   * @param {string} path - url/path to map handler to
   * @param {function} handler - an *async* function that accepts a stream object as the first argument
   */
  handle(path, handler) {
    this.router.handle(path, handler);
  }

  /**
   * Adds a middleware that always executes before handles
   * @param {function} handler - an *async* function that accepts a stream object as the first argument
   */
  use(handler) {
    this.router.use(handler);
  }

  /**
   * Bind instance to listen on a given port
   * @param {number} port - an open port
   * @param {function} callback
   */
  listen(port, callback) {
    this.server.listen(port, callback);
  }
}

module.exports = Quart;
